lib/derivation.ml

```diff
@@ -195,3 +194,32 @@ let rec derive context = function
      (* TODO: fully support map update. see: https://github.com/dwango/fialyzer/issues/102#issuecomment-461787511 *)
      derive context map >>= fun (ty_map, c_map) ->
      Ok (Type.of_elem TyAnyMap, C.Conj [c_map; C.Subtype {lhs=ty_map; rhs=Type.of_elem TyAnyMap; link=map}])
+and derive_pattern context pattern = 
+    begin match pattern with
+    | PatVar v -> 
+      begin match Context.find context (Context.Key.Var v) with
+      | Some (ty) ->
+        Ok (ty, C.Empty)
+      | None ->
+        failwith "cannot reach here"
+      end
+    | PatTuple es -> 
+      result_map_m ~f:(derive_pattern context) es
+      >>| List.unzip
+      >>| fun (tys, cs) -> (Type.of_elem (TyTuple tys), C.Conj cs)
+    | PatConstant c -> 
+      Ok (Type.of_elem (TySingleton c), C.Empty)
+    | PatCons (p1, p2) -> 
+      let alpha = new_tyvar() in
+      derive_pattern context p1 >>= fun (ty_hd, c_hd) ->
+      derive_pattern context p2 >>= fun (ty_tl, c_tl) ->
+      Ok (Type.of_elem (TyList (Type.sup alpha ty_hd)), C.Conj [
+        C.Eq {lhs=ty_tl; rhs=Type.of_elem (TyList alpha); link=ListNil (* Dummy *)};
+        c_hd;
+        c_tl
+      ])
+    | PatNil -> 
+      Ok (Type.of_elem (TyList TyBottom), C.Empty)
```

Comment:
To avoid copying, how about as follows:
```ocaml
begin match pattern with
| PatVar _ | PatTuple _ | PatConstant | PatCons _ | PatNil ->
  derive (pattern_to_expr pattern)
| PatMap ->
  ...
```